clouddriver-event/src/main/kotlin/com/netflix/spinnaker/clouddriver/event/persistence/MemoryEventRepository.kt

```diff
@@ -0,0 +1,168 @@
+import com.netflix.spinnaker.clouddriver.event.EventPublisher
+import com.netflix.spinnaker.clouddriver.event.SpinEvent
+import com.netflix.spinnaker.clouddriver.event.config.MemoryEventRepositoryConfigProperties
+import com.netflix.spinnaker.clouddriver.event.exceptions.AggregateChangeRejectedException
+import com.netflix.spinnaker.kork.exceptions.SystemException
+import org.slf4j.LoggerFactory
+import org.springframework.scheduling.annotation.Scheduled
+import java.time.Duration
+import java.time.Instant
+import java.util.concurrent.ConcurrentHashMap
+import kotlin.math.max
+
+/**
+ * An in-memory only [EventRepository]. This implementation should only be used for testing.
+ */
+class MemoryEventRepository(
+  private val config: MemoryEventRepositoryConfigProperties,
+  private val eventPublisher: EventPublisher,
+  private val registry: Registry
+) : EventRepository {
+
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+
+  private val aggregateCountId = registry.createId("eventing.aggregates")
+  private val aggregateWriteCountId = registry.createId("eventing.aggregates.writes")
+  private val aggregateReadCountId = registry.createId("eventing.aggregates.reads")
+  private val eventCountId = registry.createId("eventing.events")
+  private val eventWriteCountId = registry.createId("eventing.events.writes")
+  private val eventReadCountId = registry.createId("eventing.events.reads")
+
+  private val events: MutableMap<Aggregate, MutableList<SpinEvent>> = ConcurrentHashMap()
+
+  override fun save(aggregateType: String, aggregateId: String, originatingVersion: Long, events: List<SpinEvent>) {
+    registry.counter(aggregateWriteCountId).increment()
+
+    val aggregate = getAggregate(aggregateType, aggregateId)
+
+    if (aggregate.version != originatingVersion) {
+      // If this is being thrown, ensure that the originating process is retried on the latest aggregate version
+      // by re-reading the events list.
+      throw AggregateChangeRejectedException(
+        "Attempting to save events against an old aggregate version " +
+        "(version: ${aggregate.version}, originatingVersion: $originatingVersion)")
+    }
+
+    val currentSequence = this.events[aggregate]!!.map { it.metadata.sequence }.max() ?: 0
+    events.forEachIndexed { index, sagaEvent ->
+      // TODO(rz): Plugin more metadata (provenance, serviceVersion, etc)
+      sagaEvent.metadata = EventMetadata(
+        sequence = currentSequence + (index + 1),
+        originatingVersion = originatingVersion
+      )
+    }
+
+    registry.counter(eventWriteCountId).increment(events.size.toLong())
+    this.events[aggregate]!!.addAll(events)
+    aggregate.version = aggregate.version + 1
+
+    events.joinToString(",") { it.javaClass.simpleName }
+    log.debug("Saved $aggregateType/$aggregateId@${aggregate.version}: " +
+      "[${events.joinToString(",") { it.javaClass.simpleName }}]")
+
+    events.forEach { eventPublisher.publish(it) }
```

Comment:
given that the `eventPublisher` can be (as is now) synchronous, I think it's a slightly weird/unexpected for the caller of this method to have to wait on all publishers having to process the events being saved. I think the caller would expect the events to be saved durably but not necessarily broadcast on the same blocking operation

Reply:
This is a result of the current Saga framework design. You raise a good point, so I'm working on refactoring the Saga lib to separate the flow propagation from being event-driven to just informed by event data. This will let me remove this specific call and allow event publishing to be handled entirely asynchronously. This will also have the side effect of making `SynchronousEventPublisher` basically useless (and therefore, I'll re-evaluate the default `EventPublisher` implementation).

Thanks!